pydeseq2/dds.py

```diff
@@ -226,55 +237,57 @@ def __init__(
             )
 
         self.fit_type = fit_type
+        self.design = design
+        self.factor_storage = None
+        self.variable_to_factors = None
 
-        # Convert design_factors to list if a single string was provided.
-        self.design_factors = (
-            [design_factors] if isinstance(design_factors, str) else design_factors
-        )
-        self.continuous_factors = continuous_factors
-
-        if self.obs[self.design_factors].isna().any().any():
-            raise ValueError("NaNs are not allowed in the design factors.")
-        self.obs[self.design_factors] = self.obs[self.design_factors].astype(str)
-
-        # Check that design factors don't contain underscores. If so, convert them to
-        # hyphens.
-        if np.any(["_" in factor for factor in self.design_factors]):
+        if continuous_factors is not None:
             warnings.warn(
-                """Same factor names in the design contain underscores ('_'). They will
-                be converted to hyphens ('-').""",
-                UserWarning,
+                "continuous_factors is deprecated and will soon be removed."
+                "Continuous factors are now automatically detected from the design,"
+                "or can be cast to categorical using the C() operator in the formula",
+                DeprecationWarning,
                 stacklevel=2,
             )
 
-            new_factors = replace_underscores(self.design_factors)
-
-            self.obs.rename(
-                columns=dict(zip(self.design_factors, new_factors)),
-                inplace=True,
+        if ref_level is not None:
+            warnings.warn(
+                "ref_level is deprecated and no longer has any effect. It will be"
+                "removed in a future release.",
+                DeprecationWarning,
+                stacklevel=2,
             )
 
-            self.design_factors = new_factors
+        if design_factors is not None:
+            warnings.warn(
+                "design_factors is deprecated and will soon be removed."
+                "Please consider providing a formulaic formula using the design argument"
+                "instead.",
+                DeprecationWarning,
+                stacklevel=2,
+            )
+            design_factors = (
+                design_factors if isinstance(design_factors, list) else [design_factors]
+            )
+            self.design = "~" + " + ".join(design_factors)
 
-            # Also check continuous factors
-            if self.continuous_factors is not None:
-                self.continuous_factors = replace_underscores(self.continuous_factors)
+        assert isinstance(self.design, (str, pd.DataFrame)) or isinstance(
```

Comment:
nitpick: if this is meant as a user-facing error message, it should probably be a ValueError instead of an assertion